src/util/bson/bson.cpp

```diff
@@ -102,12 +102,12 @@ Bson& Bson::operator=(Bson&& v) noexcept {
             new (&string_val) std::string(std::move(v.string_val));
             break;
         case Type::Document:
-            if (document_val) delete document_val;
+//            if (document_val) delete document_val;
```

Comment:
Tricky, we have two cases to consider based on above: `if (m_type != v.m_type) this->~Bson();` 
1) Changing from non-document to Document type; then according to above the memory is cleaned up according to the manual destructor call and we should not again delete memory here. `if (document_val) ...` does not protect us because `document_val` coming from the union will be some random memory and not a `nullptr`.
1) Changing from Document type to again Document type; then the destructor is not called above and we leak memory here if not cleaned up.

Using a `std::unique_ptr` would have simplified this. But in lieu of changing the type, I'd suggest we don't set `m_type` until after the switch statement so that we can properly check which type we're coming from in this code and clean up memory only if necessary.